Add support for prefixing calls to HTCondor binaries to the HTCondor runner

domgz opened the pull request:
This PR is meant to be used for the HTCondor migration. It defines an optional parameter `prefix` for the HTCondor job runner that prepends the parameter value to all calls to HTCondor binaries, such as `condor_submit`, `condor_rm` or `condor_ssh_to_job`.

This enables having two HTCondor runners defined in the Galaxy job configuration file, with each one calling different executable files and thus allowing to route jobs to two different HTCondor clusters.

The shortcoming of this approach is that the contents of `prefix` are constrained to strings that, when prepended to the `condor_*` commands, still point to an executable (i.e. no shell commands). Thus, such executable files need to be created and placed somewhere. A minimal example is available [here](https://github.com/sanjaysrikakulam/esg_wp4_galaxy/compare/15f239fc47140b9f65e80b7fefd4500c8fc38e2c...kysrpex:sanjaysrikakulam-esg_wp4_galaxy:6d987e36b322c21fdabc6a15ea0695e793002e0e). This constraint exists because of how Galaxy calls the [`subprocess.Popen` constructor](https://docs.python.org/3.7/library/subprocess.html#subprocess.Popen) (using `shell=False`).

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [X] Instructions for manual testing are as follows:
  1. See [this minimal example](https://github.com/sanjaysrikakulam/esg_wp4_galaxy/compare/15f239fc47140b9f65e80b7fefd4500c8fc38e2c...kysrpex:sanjaysrikakulam-esg_wp4_galaxy:6d987e36b322c21fdabc6a15ea0695e793002e0e).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Add support for prefixing calls to HTCondor binaries to the HTCondor runner

Define an optional parameter `prefix` for the HTCondor job runner that prepends the parameter value to all calls to HTCondor binaries, such as `condor_submit`, `condor_rm` or `condor_ssh_to_job`.

domgz commented:
Do you think we should invest more time in a better solution and/or something that is really upstreamable? Or is this ok to complete the migration?

domgz commented:
@sanjaysrikakulam This is the same patch to the Galaxy code we have just tested.

bgruening commented:
Wouldn't it be better to have a overwrite per command?

docker_docker_cdocker_odocker_cdocker_c

e.g. `condor_rm_cmd` and `condor_queue_cmd`  ...

domgz commented:
> Wouldn't it be better to have a overwrite per command?
> 
> docker_docker_cdocker_odocker_cdocker_c
> 
> e.g. `condor_rm_cmd` and `condor_queue_cmd` ...

@bgruening  So the answer to

> Do you think we should invest more time in a better solution and/or something that is really upstreamable?

is yes?

bgruening commented:
Sorry I wanted to paste this link above: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/job_conf.xml.sample_advanced#L579

bgruening commented:
I don't know please discuss this internally. I assume using `condor_rm_cmd` makes the patch here and your ansible easier and cleaner.

domgz committed:
Add support for overriding HTCondor commands to the HTCondor runner

Define three optional parameters `condor_rm_cmd`, `condor_ssh_to_job_cmd`, `condor_submit_cmd` that override calls to `condor_rm`, `condor_ssh_to_job` and `condor_submit` with the parameters' values.

In addition, the runner now invokes the commands through a shell. This allows tricks such as passing the `condor_*` invocations as arguments to other commands, or performing other tasks before the calls.

domgz commented:
The [last commit](https://github.com/usegalaxy-eu/galaxy/pull/202/commits/e12e4b078c658c3731b7cd7740247c092fee19cc) adds the three optional parameters `condor_rm_cmd`, `condor_ssh_to_job_cmd`, `condor_submit_cmd` that [you requested](https://github.com/usegalaxy-eu/galaxy/pull/202#issuecomment-1787329291) (the runner does not call any other condor commands). 

In addition and more importantly, it also makes the runner invoke everything via shell, so

> The shortcoming of this approach is that the contents of `prefix` are constrained to strings that, when prepended to the `condor_*` commands, still point to an executable (i.e. no shell commands). Thus, such executable files need to be created and placed somewhere. A minimal example is available [here](https://github.com/sanjaysrikakulam/esg_wp4_galaxy/compare/15f239fc47140b9f65e80b7fefd4500c8fc38e2c...kysrpex:sanjaysrikakulam-esg_wp4_galaxy:6d987e36b322c21fdabc6a15ea0695e793002e0e). This constraint exists because of how Galaxy calls the [`subprocess.Popen` constructor](https://docs.python.org/3.7/library/subprocess.html#subprocess.Popen) (using `shell=False`).

no longer applies and one can do tricks like this (this is what I did to test it),

```shell
runners:
  condor:
    load: galaxy.jobs.runners.condor:CondorJobRunner
  condor_secondary:
    load: galaxy.jobs.runners.condor:CondorJobRunner
    prefix: "touch /tmp/prefix_works; "
    condor_rm_cmd: "touch /tmp/condor_rm_cmd_works; condor_rm"
    condor_ssh_to_job_cmd: "touch /tmp/condor_ssh_to_job_cmd_works; condor_ssh_to_job"
    condor_submit_cmd: "touch /tmp/condor_submit_cmd_works; condor_submit"
```

which definitely makes the Ansible playbooks cleaner, because then `systemd-run` can be set as `prefix`.

domgz commented:
Additionally, while testing this I found out that there is a bug in Galaxy that prevents it from stopping and killing Docker containers. Thus, whenever a user clicks the trash icon to remove a running Docker job, the container is not stopped. The error manifests in the logs as follows,

```
Nov 03 10:**:** sn06.galaxyproject.eu python[2314207]: galaxy.jobs.runners.condor WARNING 2023-11-03 10:**:**,*** [pN:handler_sn06_*,p:2314207,tN:JobHandlerStopQueue.monitor_thread] stop_job(): ********: trying to kill container failed. ('commands')
```

and [this](https://github.com/usegalaxy-eu/galaxy/blob/dd9c34c176ea30d6e3cfcbcbcfb16f17ae1d14a4/lib/galaxy/jobs/runners/condor.py#L302) is the line of code that triggers it. The reason is a `KeyError` (`"commands"` key). `cont.container_info` is an instance of `galaxy.model.custom_types.MutationDict`, but printing `cont.container_info` shows `{}` (an empty dictionary).

I will report the issue on the [Galaxy issue tracker](https://github.com/galaxyproject/galaxy/issues).

domgz committed:
Fix `condor_submit` and `condor_stop` docstrings

bgruening commented on doc/source/admin/cluster.md:
```diff
@@ -216,6 +219,20 @@ If you need to add additional parameters to your condor submission, you can do s
+The runner accepts the following optional parameters.
+
+``prefix``
+: A string that will be to prepended to HTCondor command line tool invocations. Defaults to an empty string.
```
Make clear that the prefix is followed by the command __without__ any spaces between. So you need to add a trailing space to prefix if you want that.

domgz replied:
```suggestion
: A string that will be to prepended to HTCondor command line tool invocations. Defaults to an empty string. Note that
  the runner **does not include a space** between the prefix and the command, so most likely you want to add a space
  a the end of your `prefix`.
```

I will use this text then.

bgruening commented on lib/galaxy/jobs/runners/condor.py:
```diff
@@ -52,6 +52,24 @@ class CondorJobRunner(AsynchronousJobRunner):
+        runner_param_specs = {
+            "prefix": dict(map=str, default=""),
+            # String to prepend to HTCondor command line tool invocations. Defaults to an empty string.
+            "condor_rm_cmd": dict(map=str, default="condor_rm"),
```
what are you doing here with the dict and the map?

domgz replied:
Runner parameters (`runner_params`) are an instance of [`RunnerParams`](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/jobs/runners/__init__.py#L76). `RunnerParams` inherits from [`ParamsWithSpecs`](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/util/__init__.py#L1279).

The point of `ParamsWithSpecs` is to enforce restrictions (`specs`) on some data (`params`). Have a a look [at the code](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/util/__init__.py#L1285-L1295) for greater details. In few words, it disallows keys not defined in the spec, it can process/coerce the values using the `map` function, and reject invalid values using a function `valid` that determines their validity. Specs are **not optional**, and seem to be expected to be a [`Mapping`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping).

bgruening commented on lib/galaxy/jobs/runners/util/condor/__init__.py:
```diff
@@ -67,15 +68,21 @@ def build_submit_description(executable, output, error, user_log, query_params):
 
 
-def condor_submit(submit_file):
+def condor_submit(submit_file, prefix: Optional[str] = None, command: Optional[str] = None):
```
Is there somewhere a case when you need to set prefix and command to something else than `self.runner_params.get("condor_submit_cmd")` ? If not can you not remove the optional arguments?

Or if you want to keep them, wouldn't it make more sense to do something like `prefix = prefix or self.runner_params.get("prefix")`

domgz replied:
> prefix = prefix or self.runner_params.get("prefix")

I cannot do this because `condor_submit` is not a _bound method_ and thus there is no `self` in its scope.

> Is there somewhere a case when you need to set prefix and command to something else than `self.runner_params.get("condor_submit_cmd")` ? If not can you not remove the optional arguments?

No, but this is not an arbitrary decision nevertheless. If you really want to enter in this discussion we can, but I really do not think it is worth the debate. I think we should rather be keeping our heads busy with, for example, how is the monitoring supposed to be going to work during the migration.

bgruening commented:
Have you read https://docs.python.org/3.7/library/subprocess.html#security-considerations and why we should not use shell=True? Are you sure you can not get the same results with shell=False?

domgz commented:
> Have you read https://docs.python.org/3.7/library/subprocess.html#security-considerations and why we should not use shell=True? Are you sure you can not get the same results with shell=False?

That's a very good point. Let's look at what are the possible ways to take advantage of shell injection here:
- [condor_submit](https://github.com/kysrpex/galaxy/blob/3f99db0e58cbcffee147cb24df61a54bc3dff87c/lib/galaxy/jobs/runners/util/condor/__init__.py#L85): `{prefix}{command} {submit_file}`. We configure the prefix and the command. Can `submit_file` be malicious? That's a [string with a path on it](https://github.com/kysrpex/galaxy/blob/3f99db0e58cbcffee147cb24df61a54bc3dff87c/lib/galaxy/jobs/runners/condor.py#L108), [crafted by Galaxy](https://github.com/kysrpex/galaxy/blob/3f99db0e58cbcffee147cb24df61a54bc3dff87c/lib/galaxy/jobs/runners/condor.py#L85C45-L85C45).
- [condor_rm](https://github.com/kysrpex/galaxy/blob/3f99db0e58cbcffee147cb24df61a54bc3dff87c/lib/galaxy/jobs/runners/util/condor/__init__.py#L108): `{prefix}{command} {external_id}`. `external_id` is generated by HTCondor.
- [condor_ssh_to_job](https://github.com/kysrpex/galaxy/blob/3f99db0e58cbcffee147cb24df61a54bc3dff87c/lib/galaxy/jobs/runners/condor.py#L331-L338): `f"{self.runner_params.get('prefix', '')}{self.runner_params.get('condor_ssh_to_job_cmd')} {external_job_id} {command}`. Again, `external_job_id` is generated by HTCondor. `command` comes from `job_wrapper.get_job().container.container_info["commands"][command]`. `job_wrapper.get_job()` yields a `model.Job`, so I guess `container` is what is stored on [the job container association table](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/model/__init__.py#L1341). Looking at `psql` that seems to be a JSON blob. I have no idea what could be in there. I would have to spend a few hours diving into the Galaxy code to find out with certainty, or alternatively (although less reliably), reading examples from the database.

I think the conclusion is, if we trust Galaxy developers and HTCondor developers, then the two first points are ok. If we trust the tools we install, then the third point is probably also ok. We are not dealing with untrusted inputs here.

domgz committed:
Update HTCondor runner documentation

Make it clear that the `prefix` is followed by the command without any spaces in between, and thus adding a trailing space to `prefix` is recommended.

domgz commented:
If there are no further objections, let's move on?

bgruening commented:
Go for it ... but please try to get those changes or similar ones also upstream for the next release.
